### PR TITLE
allow IOBuf::append_user_data_with_meta to specify the lkey of rdma memory region

### PR DESCRIPTION
**Pull request by Tuvie**
### What problem does this PR solve?

Issue Number: [1995](https://github.com/apache/incubator-brpc/issues/1995)

Problem Summary: When using IOBuf::append_user_data, if the application specify a buffer with an offset over the base address registered with rdma::RegisterMemoryForRdma, brpc reports a failure for invalid memory region.

### What is changed and the side effects?

Changed: Add a new function named append_user_data_with_meta in IOBuf. Allow application specifying the lkey as meta to explicitly inform brpc. This removes the necessity to lookup the lkey for the IOBuf Block in rpc processing.

Side effects:
- Performance effects(性能影响): 

- Breaking backward compatibility(向后兼容性): To store the meta/lkey in IOBuf Block, the portal_next field is reused when flags is non-0. It is acceptable because portal_next is only used for tls cache of IOBuf, which does not work for user defined Blocks.

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

**Comment by wwbmmm** [comment]
`append_user_data_with_meta` not found in files changed

File: src/butil/iobuf.h [hunk 1]
```diff
@@ -246,7 +247,18 @@ friend class IOBufCutter;
     // Append the user-data to back side WITHOUT copying.
     // The user-data can be split and shared by smaller IOBufs and will be
     // deleted using the deleter func when no IOBuf references it anymore.
-    int append_user_data(void* data, size_t size, void (*deleter)(void*));
+    inline int append_user_data(void* data, size_t size, void (*deleter)(void*)) {
```
Comment (wwbmmm) [inline src/butil/iobuf.h:250]: This method should be defined in `iobuf_inl.h`

File: src/butil/iobuf.cpp [hunk 2]
```diff
@@ -1219,11 +1227,23 @@ int IOBuf::append_user_data(void* data, size_t size, void (*deleter)(void*)) {
...
     _move_back_ref(r);
     return 0;
 }
 
+uint64_t IOBuf::get_first_data_meta() {
+    if (_ref_num() == 0) {
+        return 0;
+    }
+    IOBuf::BlockRef const& r = _ref_at(0);
+    if (!r.block->flags) {
```
Comment (wwbmmm) [inline src/butil/iobuf.cpp:1241]: flags & IOBUF_BLOCK_FLAGS_USER_DATA

File: src/butil/iobuf.h [hunk 3]
```diff
@@ -66,6 +66,7 @@ friend class IOBufCutter;
 public:
     static const size_t DEFAULT_BLOCK_SIZE = 8192;
     static const size_t INITIAL_CAP = 32; // must be power of 2
+    static const size_t BLOCK_META_OFFSET = 16;
```
Comment (wwbmmm) [inline src/butil/iobuf.h:69]: Where is it used?

File: test/brpc_rdma_unittest.cpp [hunk 4]
```diff
@@ -80,6 +80,7 @@ extern bool g_skip_rdma_init;
 }
 }
 
+static const size_t MAX_USER_MRS = 16;
```
Comment (wwbmmm) [inline test/brpc_rdma_unittest.cpp:83]: Where is it used?

Comment (Tuvie) [reply test/brpc_rdma_unittest.cpp]: should remove

File: src/butil/iobuf.cpp (hunk 2 above)
Comment (Tuvie) [reply src/butil/iobuf.cpp]: OK

File: src/butil/iobuf.h (hunk 1 above)
Comment (Tuvie) [reply src/butil/iobuf.h]: OK

File: src/butil/iobuf.cpp [hunk 5]
```diff
@@ -1206,7 +1211,14 @@ int IOBuf::appendv(const const_iovec* vec, size_t n) {
...
 }
 
-int IOBuf::append_user_data(void* data, size_t size, void (*deleter)(void*)) {
+int IOBuf::append_user_data_with_meta(void* data,
+                                      size_t size,
+                                      void (*deleter)(void*),
+                                      uint64_t meta) {
+    if (size == 0) {
+        LOG(WARNING) << "data_size should not be 0";
+        return -1;
```
Comment (372046933) [inline src/butil/iobuf.cpp:1220]: Did `size == 0` have any side effect? If not, can we return 0 here?

**Comment by wwbmmm** [comment]
LGTM

File: src/butil/iobuf.cpp (hunk 5 above)
Comment (Tuvie) [reply src/butil/iobuf.cpp:1220]: This should be unexpected. return -1 is a better choice.

**Comment by Tuvie** [comment]
@wwbmmm is this PR qualified to be merged? It seems that the Travis often fails due to some irrelevant reason.

File: src/butil/iobuf.cpp (hunk 5 above)
Comment (372046933) [reply src/butil/iobuf.cpp:1220]: @Tuvie Return 0 has been merged into master
https://github.com/apache/incubator-brpc/blob/b788325528d886639e66aa0d46c571c163ec959a/src/butil/iobuf.cpp#L1214-L1220

**Comment by Tuvie** [comment]
Has resolved conflict. Any plan to merge?

**Pull request merged**